### PR TITLE
UINV-126 Download batch vouchers

### PR DESCRIPTION
**Pull request by AndreiShumski2**
## Purpose
https://issues.folio.org/browse/UINV-126

#### TODOS and Open Questions
Could be merged after [MODINVOICE-132](https://github.com/folio-org/mod-invoice/pull/123)

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.
- [x] I've added appropriate record to the CHANGELOG.md
- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [x] If any API-related changes - okapi interfaces and permissions are reviewed/changed correspondingly
  - [x] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import { stripesConnect } from '@folio/stripes/core';
+
+import { batchVouchersFromPropResource } from '../../../common/resources';
+import BatchVouchers from './BatchVouchers';
+
+const BatchVouchersContainer = ({ mutator, batchVoucherId }) => {
+  const downloadBatchVouchers = useCallback(
+    () => {
+      mutator.batchVouchers.GET();
+      // then download file
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js:13]: but the api return json or xml, it wouldn't work with regular get

File: src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  batchVouchers: batchVouchersFromPropResource,
+});
+
+BatchVouchersContainer.propTypes = {
+  mutator: PropTypes.object.isRequired,
+  batchVoucherId: PropTypes.string,
+};
+
+BatchVouchersContainer.defaultProps = {
+  batchVoucherId: '',
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js:35]: what is the reason to put such default props? I don't see any

File: src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    () => {
+      mutator.batchVouchers.GET();
+      // then download file
+    },
+    [mutator.batchVouchers],
+  );
+
+  return (
+    <BatchVouchers
+      downloadBatchVouchers={batchVoucherId ? downloadBatchVouchers : undefined}
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js:20]: IMHO, you can follow the common pattern and put it above, something like `if (!batchVoucherId ) return null;` or make ternary in return itself

File: src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import { batchVouchersFromPropResource } from '../../../common/resources';
+import BatchVouchers from './BatchVouchers';
+
+const BatchVouchersContainer = ({ mutator, batchVoucherId }) => {
+  const downloadBatchVouchers = useCallback(
+    () => {
+      mutator.batchVouchers.GET();
+      // then download file
+    },
+    [mutator.batchVouchers],
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js:15]: don't depend on mutators

File: src/settings/BatchGroupConfigurationSettings/BatchVoucherExportsList.js [hunk 5]
```diff
@@ -25,13 +22,9 @@ const cellFormatters = {
...
-  exportButton: () => (
-    <Layout element="span" className="full textRight">
-      <IconButton
-        data-test-batch-voucher-export-download
-        icon="download"
-      />
-    </Layout>
+  // eslint-disable-next-line react/prop-types
+  exportButton: ({ batchVoucherId }) => (
+    <BatchVouchersContainer batchVoucherId={batchVoucherId} />
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchVoucherExportsList.js:27]: It's weird component name. If you want to hide button with icon and callback - maybe it makes sense to call component like `ExportVoucherButton` or something

File: src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchers.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import { IconButton } from '@folio/stripes/components';
+
+const BatchVouchers = ({ downloadBatchVouchers }) => (
+  <IconButton
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchers.js:7]: so, you component is just an usage of `<IconButton />`? IMHO, it looks like over-engineering.

File: src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js (hunk 1 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchVouchers/BatchVouchersContainer.js]: added logic

File: src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+const ExportVoucherButton = ({ batchVoucherId, format, stripes }) => {
+  const showCallout = useShowCallout();
+  const downloadBatchVouchers = useCallback(
+    async () => {
+      try {
+        const httpHeaders = {
+          'X-Okapi-Tenant': stripes.okapi.tenant,
+          'X-Okapi-Token': stripes.store.getState().okapi.token,
+          'Accept': format.toLowerCase(),
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js:22]: IMHO, it's a bad practice. Maybe it's better to create a map, where keys will be constants of batch export config format, and keys - corresponding HTTP headers

File: src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+import { BATCH_VOUCHERS_API } from '../../../common/constants';
+
+const ExportVoucherButton = ({ batchVoucherId, format, stripes }) => {
+  const showCallout = useShowCallout();
+  const downloadBatchVouchers = useCallback(
+    async () => {
+      try {
+        const httpHeaders = {
+          'X-Okapi-Tenant': stripes.okapi.tenant,
```
Comment (NikitaSedyx) [inline src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js:20]: as I remember there is a hook for okapi

File: src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js (hunk 7 above)
Comment (NikitaSedyx) [reply src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js:22]: imo it's ok to use array with all possible formats (xml and json here?)

File: src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js (hunk 8 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js:20]: found only one in [data-import](https://github.com/folio-org/ui-data-import/blob/61db482c74a3b4cb1a6274138a537878fa64ea35/src/utils/createOkapiHeaders.js#L7)

File: src/settings/BatchGroupConfigurationSettings/constants.js [hunk 9]
```diff
@@ -29,3 +29,8 @@ export const SCHEDULE_EXPORT_OPTIONS = Object.values(SCHEDULE_EXPORT).map(schedu
 }));
 
 export const RESULT_COUNT_INCREMENT = 10;
+
+export const EXPORT_FORMATS_HEADER_MAP = Object.values(EXPORT_FORMAT).reduce((acc, format) => ({
+  ...acc,
+  [format]: { 'Accept': format.toLowerCase() },
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/constants.js:35]: But this doesn't change anything - you still have coupled constants for export config format and http headers. Please revisit my initial comment, where I asked to decouple them

File: src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+const ExportVoucherButton = ({ batchVoucherId, format, stripes }) => {
+  const showCallout = useShowCallout();
+  const downloadBatchVouchers = useCallback(
+    async () => {
+      try {
+        const httpHeaders = {
+          ...EXPORT_FORMATS_HEADER_MAP[format],
+          'X-Okapi-Tenant': stripes.okapi.tenant,
+          'X-Okapi-Token': stripes.store.getState().okapi.token,
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js:23]: okapi comes as a prop, no need to do this magic

File: src/settings/BatchGroupConfigurationSettings/constants.js (hunk 9 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/constants.js]: fixed

File: src/settings/BatchGroupConfigurationSettings/BatchVoucherExportsList.js [hunk 11]
```diff
@@ -2,43 +2,39 @@ import React from 'react';
...
-    </Layout>
-  ),
-  status: ({ status }) => BATCH_VOUCHER_EXPORT_STATUS_LABEL[status],
-};
-
 const rowProps = { alignLastColToEnd: true };
 
-export function BatchVoucherExportsList({ batchVoucherExports, onNeedMoreData, recordsCount }) {
+export function BatchVoucherExportsList({ batchVoucherExports, format, onNeedMoreData, recordsCount }) {
+  const cellFormatters = {
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchVoucherExportsList.js:24]: now ot will be re-created every render, could it be the previous way?

File: src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  );
+};
+
+ExportVoucherButton.propTypes = {
+  batchVoucherId: PropTypes.string,
+  format: PropTypes.string,
+  stripes: stripesShape,
+};
+
+export default stripesConnect(ExportVoucherButton);
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/ExportVoucherButton/ExportVoucherButton.js:68]: IMHO, you don't need stripesConnect here. have you checked its code? Maybe withStripes is enough

**Pull request merged**